### PR TITLE
/user command which gives user details

### PR DESCRIPTION
**Pull request by j24m**
## Description
- This pull request introduces a new discord slash command **/user** which takes as an argument username and provides details of that user.
- Issue Ticket :  https://github.com/Real-Dev-Squad/discord-slash-commands/issues/106
- When this command is executed, it returns the following details of the user :
 
```md
##  Jyotsna's Details

   **Full Name :** Jyotsna Mehta
   **RDS Discord Joined At :** 16 June 2023, 9:20 PM IST
   **State :** IDLE
```
image

### CODE REVIEW

File: src/controllers/userCommand.ts [hunk 1]
```diff
@@ -0,0 +1,14 @@
+import { env } from "../typeDefinitions/default.types";
+import {} from "../typeDefinitions/discordMessage.types";
```
Comment (sahsisunny) [inline src/controllers/userCommand.ts:2]: What are you importing here from `discordMessage.types` file?

File: src/typeDefinitions/userBackend.types.d.ts [hunk 2]
```diff
@@ -2,12 +2,19 @@ export interface UserBackend {
...
+    discordId: string;
+    github_display_name: string;
+    github_id: string;
     username: string;
     first_name: string;
     last_name: string;
-    github_id: string;
-    discordId: string;
-    github_display_name: string;
-    isMember: boolean;
```
Comment (sahsisunny) [inline src/typeDefinitions/userBackend.types.d.ts:11]: Please don't make unnecessary changes.

File: src/typeDefinitions/userBackend.types.d.ts [hunk 3]
```diff
@@ -2,12 +2,19 @@ export interface UserBackend {
...
+    roles: {
+      in_discord: boolean;
+      super_user: boolean;
+      archived: boolean;
+    };
+    incompleteUserDetails: boolean;
+    state: string;
```
Comment (sahsisunny) [inline src/typeDefinitions/userBackend.types.d.ts:18]: Can you make it optional?

File: src/utils/getUserDetails.ts [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+      `${RDS_BASE_API_URL}/users/?dev=true&discordId=${discordId}`
+    );
+
+    if (!response.ok) {
+      throw new Error(`Failed to fetch user details for ${discordId}.`);
+    }
+
+    const data = await response.json();
+    const user = `${JSON.stringify(data)}`;
+    return formatUserDetails(user);
```
Comment (sahsisunny) [inline src/utils/getUserDetails.ts:16]: Instead of returning formatted text, can you return only the API response so we can reuse the data?

File: src/utils/registerCommands.ts [hunk 5]
```diff
@@ -21,9 +21,9 @@ export async function registerCommands(
       method: "PUT",
       body: JSON.stringify(commands),
     });
-
-    if (response.ok) console.log("Registered all commands");
-    else console.error("Error Registering Commands");
+    if (response.ok) {
+      console.log("Registered all commands");
+    } else console.error("Error Registering Commands");
```
Comment (sahsisunny) [inline src/utils/registerCommands.ts:26]: ?

File: tests/unit/utils/getUserDetails.test.ts [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+        state: "Active",
+      },
+    };
+
+    jest
+      .spyOn(global, "fetch")
+      .mockImplementation(() => Promise.resolve(new JSONResponse(userData)));
+
+    const userId = "123456";
+    const result = await getUserDetails(userId);
```
Comment (shubhamsigdar1) [inline tests/unit/utils/getUserDetails.test.ts:27]: Nit: you are taking userId or discordId as a params

File: tests/fixtures/fixture.ts [hunk 7]
```diff
@@ -160,14 +160,21 @@ export const discordUserData = {
...
+    username: "jane",
+    first_name: "Jane",
+    last_name: "Doe",
+    roles: {
+      in_discord: true,
+      super_user: false,
+      archived: false,
+    },
+    incompleteUserDetails: false,
+    state: "IDLE",
```
Comment (sahsisunny) [inline tests/fixtures/fixture.ts:177]: To ensure that the fixture data matches the original API response data, please make the necessary corrections, and double-check the data types used for fields.

File: tests/unit/utils/getUserDetails.test.ts [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+    const userData = {
+      user: {
+        discordId: "123456",
+        first_name: "John",
+        last_name: "Doe",
+        discordJoinedAt: "2023-09-13T09:30:00.000Z",
+        state: "Active",
+      },
+    };
+
```
Comment (sahsisunny) [inline tests/unit/utils/getUserDetails.test.ts:21]: Please use fixture data for testing or move this dummy data to the fixtures file.

File: tests/unit/utils/formatUserDetails.test.ts [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+    const user = JSON.stringify({
+      user: {
+        first_name: "John",
+        last_name: "Doe",
+        discordJoinedAt: "2023-09-13T09:30:00.000Z",
+        state: "Active",
+      },
+    });
+
+    const formattedDetails = formatUserDetails(user).trim();
```
Comment (shubhamsigdar1) [inline tests/unit/utils/formatUserDetails.test.ts:14]: Nit: why used trim?

**Review by shubhamsigdar1 — COMMENTED** [review]
great work Jyotsna
Just go through sunny-s comment
and rest look good 
ping me after changes I will approve

File: tests/unit/utils/getUserDetails.test.ts (hunk 6 above)
Comment (j24m) [reply tests/unit/utils/getUserDetails.test.ts:27]: Eventhough it is userId here, but internally it will take discordId.

File: tests/unit/utils/formatUserDetails.test.ts (hunk 9 above)
Comment (j24m) [reply tests/unit/utils/formatUserDetails.test.ts:14]: In order to remove any white spaces before and after the string so that the expected test case matches with the resulting data.

File: src/controllers/baseHandler.ts [hunk 10]
```diff
@@ -63,6 +66,11 @@ export async function baseHandler(
       return await mentionEachUser(transformedArgument, env);
     }
 
+    case getCommandName(USER): {
+      const data = message.data?.options as Array<messageRequestDataOptions>;
+      return await userCommand(data[0].value, env);
```
Comment (iamitprakash) [inline src/controllers/baseHandler.ts:71]: what if data is null

File: src/controllers/userCommand.ts [hunk 11]
```diff
@@ -0,0 +1,14 @@
...
+import { discordTextResponse } from "../utils/discordResponse";
+import { getUserDetails } from "../utils/getUserDetails";
+
+export async function userCommand(discordId: string, env: env) {
+  try {
+    const UserDetails = await getUserDetails(discordId);
+    return discordTextResponse(UserDetails);
+  } catch (error) {
+    console.error("User not found", error);
```
Comment (iamitprakash) [inline src/controllers/userCommand.ts:11]: why console

File: src/utils/getUserDetails.ts [hunk 12]
```diff
@@ -0,0 +1,21 @@
...
+  } catch (error) {
+    console.error("An error occurred while fetching user details:", error);
```
Comment (iamitprakash) [inline src/utils/getUserDetails.ts:18]: why console

File: src/utils/registerCommands.ts [hunk 13]
```diff
@@ -21,9 +21,9 @@ export async function registerCommands(
       method: "PUT",
       body: JSON.stringify(commands),
     });
-
-    if (response.ok) console.log("Registered all commands");
-    else console.error("Error Registering Commands");
+    if (response.ok) {
+      console.log("Registered all commands");
```
Comment (iamitprakash) [inline src/utils/registerCommands.ts:25]: why console

File: src/constants/commands.ts [hunk 14]
```diff
@@ -28,6 +28,19 @@ export const MENTION_EACH = {
...
 };
 
+export const USER = {
+  name: "user",
+  description: "Replies with specific user details in the channel",
+  options: [
+    {
+      name: "username",
+      description: "Takes username and shows user details",
+      type: 6, // 2 is type SUB_COMMAND_GROUP
```
Comment (iamitprakash) [inline src/constants/commands.ts:38]: why type is hard coded

**Review by iamitprakash — CHANGES_REQUESTED** [review]

File: src/constants/commands.ts (hunk 14 above)
Comment (j24m) [reply src/constants/commands.ts:38]: Mandatory field with respect to the command I have built.
Reference : 
- https://discord.com/developers/docs/interactions/application-commands
image

File: src/utils/registerCommands.ts (hunk 13 above)
Comment (j24m) [reply src/utils/registerCommands.ts:25]: When we run `npm run register` to register the command we built, we need an acknowledgement whether the command was registered successfully or not.

File: src/controllers/baseHandler.ts (hunk 10 above)
Comment (j24m) [reply src/controllers/baseHandler.ts:71]: Data(username) will never be null because the /user command takes an argument username which always shows name suggestions for people who exist in the server. 
**Reference Image :** 
image

**Review by gauravsinhaweb — COMMENTED** [review]
Looks good to me, resolve above comments by sunny and tejas, let me know for approval

File: tests/fixtures/fixture.ts (hunk 7 above)
Comment (j24m) [reply tests/fixtures/fixture.ts:177]: Crosschecked the format in which data is being fetched. Format is correct.

File: src/controllers/userCommand.ts (hunk 11 above)
Comment (j24m) [reply src/controllers/userCommand.ts:11]: Removed.

File: src/utils/getUserDetails.ts (hunk 12 above)
Comment (j24m) [reply src/utils/getUserDetails.ts:18]: Removed.

File: tests/unit/utils/getUserDetails.test.ts (hunk 8 above)
Comment (j24m) [reply tests/unit/utils/getUserDetails.test.ts:21]: Done

**Review by iamitprakash — CHANGES_REQUESTED** [review]
"IST" can we remove this hard coded value

**Comment by iamitprakash** [comment]
@shubhamsigdar1  what about IST hard coded value

**Review by iamitprakash — CHANGES_REQUESTED** [review]
Please add After Before test case stats
